alf/utils/gin_utils.py

```diff
@@ -53,3 +56,91 @@ def inoperative_config_str(max_line_length=80, continuation_indent=4):
     gin.config._OPERATIVE_CONFIG = operative_config
     gin.config._IMPORTED_MODULES = imported_module
     return inoperative_str
+
+
+def _config_gin_eval(func):
+    """Decorator to make function `gin_eval` configurable.
+    And it get and pass expected frame context variables `globals` and `locals`
+    to `func` when is actually called. See `gin_eval` for more details.
+    """
+
+    name = func.__name__
+    module = getattr(func, '__module__', None)
+    selector = module + '.' + name if module else name
+    fn = gin.config._ensure_wrappability(func)
+
+    @six.wraps(fn)
+    def gin_wrapper(**kwargs):
+        scope_components = gin.config.current_scope()
+        new_kwargs = {}
+        for i in range(len(scope_components) + 1):
+            partial_scope_str = '/'.join(scope_components[:i])
+            new_kwargs.update(
+                gin.config._CONFIG.get((partial_scope_str, selector), {}))
+        scope_str = partial_scope_str
+        operative_parameter_values = {}
+        operative_parameter_values.update(new_kwargs)
+        gin.config._OPERATIVE_CONFIG.setdefault(
+            (scope_str, selector), {}).update(operative_parameter_values)
+        new_kwargs = copy.deepcopy(new_kwargs)
+        # hack here ,injection was actually called at frame_trace[-7]
+        # (gin-config>=0.1.3,<=0.3.3)
+        frame = sys._getframe(7)
+        context = lambda c: (c, frame.f_globals, frame.f_locals)
+        new_kwargs = {k: context(v) for k, v, in new_kwargs.items()}
+        new_kwargs.update(kwargs)
+        return fn(**new_kwargs)
+
+    gin.config._REGISTRY[selector] = gin.config.Configurable(
+        gin_wrapper,
+        name=name,
+        module=module,
+        whitelist=None,
+        blacklist=None,
+        selector=selector)
+    return gin_wrapper
+
+
+@_config_gin_eval
+def gin_eval(source):
```

Comment:
Now I actually think `source` would be better called `str`. It's shorter and can indicate the type of the argument. What do you think?

Reply:
renamed

Reply:
actually i just follow the signature of builtin function `eval(__source, __globals, __locals)`